jsign-core/src/main/java/net/jsign/mscab/MSCabFile.java

```diff
@@ -0,0 +1,487 @@
+            byte singleChar;
+            ByteBuffer bbtmp = ByteBuffer.allocate(1);
+            do {
+                bbtmp.clear();
+                bbtmp.limit(1);
+                channel.read(bbtmp);
+                bbtmp.flip();
+                singleChar = bbtmp.array()[0];
+                bos.write(singleChar);
+            } while (singleChar != 0);
+            return bos.toByteArray();
+        }
+    }
+
+    @Override
+    public synchronized byte[] computeDigest(MessageDigest digest) throws IOException {
+        CFHeader modifiedHeader = new CFHeader(header);
+        if (!CFHeaderFlag.RESERVE_PRESENT.checkFrom(header.flags)) {
+            ByteBuffer bb = ByteBuffer.allocate(20)
+                    .order(ByteOrder.LITTLE_ENDIAN);
+
+            modifiedHeader.cbCFHeader = 20;
+            modifiedHeader.cbCabinet += 24;
+            modifiedHeader.coffFiles += 24;
+            modifiedHeader.flags |= CFHeaderFlag.RESERVE_PRESENT.getValue();
+
+            bb.putInt(0x00100000);
+            bb.putInt((int)modifiedHeader.cbCabinet); // size of cab file
+            bb.putInt(0); // size of signature asn1 der
+            bb.putInt(0);
+            bb.putInt(0);
+
+            modifiedHeader.abReserved = bb.array();
+        }
+        modifiedHeader.headerDigestUpdate(digest);
+
+        ByteBuffer bbtmp = ByteBuffer.allocate(4096).order(ByteOrder.LITTLE_ENDIAN);
+        int off = header.getHeaderSize();
+        channel.position(off);
+
+        if (CFHeaderFlag.PREV_CABINET.checkFrom(header.flags)) {
+            byte[] szCabinetPrev = readNullTerminatedString(channel);
+            byte[] szDiskPrev = readNullTerminatedString(channel);
+            digest.update(szCabinetPrev);
+            digest.update(szDiskPrev);
+            off += szCabinetPrev.length + szDiskPrev.length;
+        }
+
+        if (CFHeaderFlag.NEXT_CABINET.checkFrom(header.flags)) {
+            byte[] szCabinetNext = readNullTerminatedString(channel);
+            byte[] szDiskNext = readNullTerminatedString(channel);
+            digest.update(szCabinetNext);
+            digest.update(szDiskNext);
+            off += szCabinetNext.length + szDiskNext.length;
+        }
+
+        for (int i = 0; i < header.cFolders; i++) {
+            bbtmp.clear();
+            bbtmp.limit(8);
+            channel.read(bbtmp);
+            bbtmp.flip();
+            if (!CFHeaderFlag.RESERVE_PRESENT.checkFrom(header.flags)) {
+                int a = bbtmp.getInt();
+                int b = bbtmp.getInt();
+                a += 24;
+                bbtmp.clear();
+                bbtmp.putInt(a);
+                bbtmp.putInt(b);
+                digest.update(bbtmp.array(), 0, 8);
+            } else {
+                digest.update(bbtmp.array(), 0, 8);
+            }
+            off += 8;
+        }
+
+        long endPosition = header.hasSignature() ? header.getSigPos() : channel.size();
```

Comment:
I think the value of sigpos can't be trusted to be right, an invalid cabinet could be crafted with a sigpos higher than the file size, and this would result in an infinite loop. There's probably an opportunity here to share the code computing the hash of a portion of the file with PEFile (specifically the `updateDigest` method). I'll take care of that.